packages/expo-cli/src/commands/build/ios/IOSBuilder.ts

```diff
@@ -144,21 +145,102 @@ See https://docs.expo.io/versions/latest/distribution/building-standalone-apps/#
+      throw e;
+    } finally {
+      const credentials = await context.ios.getAllCredentials();
+      displayProjectCredentials(experienceName, bundleIdentifier, credentials);
+    }
+  }
+
+  async _setupDistCert(
+    ctx: Context,
+    experienceName: string,
+    bundleIdentifier: string,
+    appCredentials: IosAppCredentials
+  ): Promise<void> {
+    try {
+      const nonInteractive = this.options.parent && this.options.parent.nonInteractive;
+      const distCertFromParams = await getDistCertFromParams(this.options);
+      if (distCertFromParams) {
+        await useDistCertFromParams(ctx, appCredentials, distCertFromParams);
       } else {
-        log(
-          chalk.bold.red(
-            'Failed to prepare all credentials. \nThe next time you build, we will automatically use the following configuration:'
-          )
+        await runCredentialsManager(
+          ctx,
+          new SetupIosDist({ experienceName, bundleIdentifier, nonInteractive })
         );
       }
+    } catch (e) {
+      log.error('Failed to set up Distribution Certificate');
+      throw e;
     }
+  }
 
-    const credentials = await context.ios.getAllCredentials();
-    displayProjectCredentials(experienceName, bundleIdentifier, credentials);
+  async _setupPushCert(
+    ctx: Context,
+    experienceName: string,
+    bundleIdentifier: string,
+    appCredentials: IosAppCredentials
+  ): Promise<void> {
+    try {
+      const nonInteractive = this.options.parent && this.options.parent.nonInteractive;
+      const pushKeyFromParams = await getPushKeyFromParams(this.options);
+      if (pushKeyFromParams) {
+        await usePushKeyFromParams(ctx, appCredentials, pushKeyFromParams);
+      } else {
+        await runCredentialsManager(
+          ctx,
+          new SetupIosPush({ experienceName, bundleIdentifier, nonInteractive })
+        );
+      }
+    } catch (e) {
+      log.error('Failed to set up Push Key');
+      throw e;
+    }
+  }
+
+  async _setupProvisioningProfile(
+    ctx: Context,
+    experienceName: string,
+    bundleIdentifier: string,
+    appCredentials: IosAppCredentials,
+    distributionCert: IosDistCredentials
+  ) {
+    try {
+      const nonInteractive = this.options.parent && this.options.parent.nonInteractive;
+      const provisioningProfileFromParams = await getProvisioningProfileFromParams(this.options);
+      if (provisioningProfileFromParams) {
+        await useProvisioningProfileFromParams(
+          ctx,
+          appCredentials,
+          this.options.teamId!,
+          provisioningProfileFromParams,
+          distributionCert
+        );
+      } else {
+        await runCredentialsManager(
+          ctx,
+          new SetupIosProvisioningProfile({
+            experienceName,
+            bundleIdentifier,
+            distCert: distributionCert,
+            nonInteractive,
+          })
+        );
+      }
+    } catch (e) {
+      log.error('Failed to set up Provisioning Profile');
```

Comment:
similar to above, is there any more context we can provide here? maybe a expo.fyi link that explains why this can happen, for example

Reply:
see above comment :)